repl: add reverse search

BridgeAR opened the pull request:
This adds a reverse search similar to zsh. Please check the commit messages for details.

<s>This is still WIP. It needs tests and one line length issue must be fixed.</s>
Update: done

See it in action: https://asciinema.org/a/shV3YOFu74BcBakJcvY4USNqv

Supersedes https://github.com/nodejs/node/pull/24335

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

Fishrock123 commented on lib/internal/repl/utils.js:
```diff
@@ -392,8 +400,223 @@ function setupPreview(repl, contextSymbol, bufferSymbol, active) {
+
+  const alreadyMatched = new Set();
+  const labels = {
+    r: 'bck-i-search: ',
```
Can these prompts be less abbreviated?

BridgeAR replied:
Is it OK to land this as is? We can also change the name later on in a patch.

Fishrock123 commented:
This appears to work as advertised. 👍

BridgeAR commented on lib/internal/repl/utils.js:
```diff
@@ -392,8 +400,223 @@ function setupPreview(repl, contextSymbol, bufferSymbol, active) {
 
+function setupReverseSearch(repl) {
+  // Simple terminals can't use reverse search.
+  if (process.env.TERM === 'dumb') {
```
So far we only support "dumb" and I would like to keep this consistent. We also did not document anything about it so far. It's probably not a bad idea to that in general though in `readline`. It does seem out of scope of this PR though.

BridgeAR commented:
I just addressed the nits, fixed one very unlikely issue (search lines exceeding the current terminal columns) and added a big TODO entry. Resizing the terminal will hide the overlay and fixing that requires significant work inside of readline. This should definitely not be a blocker though.

@benjamingr in what way are you uncertain that we should support reverse-i-search? It is one of my biggest missed features :D (I also plan on adding more features soonish).

Trott commented:
documentation update checkbox is checked, but no doc files have been updated I don't think?

BridgeAR commented:
@Trott I added a specific section that documents the behavior. We should probably revive https://github.com/nodejs/node/pull/20825. Not all key bindings are actually fully implemented but so far we lack any key binding documentation.

antsmartian commented on lib/internal/repl/utils.js:
```diff
@@ -392,8 +400,242 @@ function setupPreview(repl, contextSymbol, bufferSymbol, active) {
+  const alreadyMatched = new Set();
+  const labels = {
+    r: 'bck-i-search: ',
+    s: 'fwd-i-search: '
```
nit: Do we need to actually change the labels on key press? I think it might not be needed. ( IMHO, the change while I type seems to be little bit annoying)

BridgeAR replied:
I am uncertain in what way the labels would change during key presses? It is intentional to indicate the search direction, depending on what the user wants: `<ctrl> + R` for the backwards search and `<ctrl> + S` to search forwards.

antsmartian replied:
Yes. I wanted to just keep the label as `search-text` instead. Also changing the label like `failed-fwd-i-search` while I type seems to be bit annoying for the user (visually since the text appears and disappears). Again this is not a blocking comment, just my personal opinion.

BridgeAR replied:
It provides extra information to the user and that should be very valuable. 

`failed-*-i-search` will only come up in case there's no match and it stays there until the search is either cancelled or the input changes to something that matches. Not indicating that there is no match is IMO not a good idea.

Functionality wise it's almost a 1-1 copy of the ZSH reverse-i-search and ZSH is likely the most intuitive and user friendly shell out there (even apple uses it by default by now).

antsmartian replied:
I'm ok to land as it is :) when I played around it was a bit annoying to me, definitely not a blocking comment.

BridgeAR commented:
Landed in a03458396b35c455f01a7429456fd8f0b5d7a4f4...b45eeee01741013c4a37887817c71b779d9edb12 :tada:
